### ISSUE TITLE
Better support for running multiple nodes of the same type on same system

### ISSUE THREAD

**Issue by vstax** [leo-project/leofs#792]
This is related to tickets #790 and #791 and might require implementation that depends on them.

**Use case**: wish to run REST and S3 gateways on the same server, which is currently really annoying because you'll need to *copy* leo_gateway directory to e.g. "leo_rest_gateway" after each package update and place different config files there, and so on. **Another use case**: wish to run two clusters (two sets of storage nodes) on the same hardware, something that I want to do because of very different requirements (one is very large, objects are never updated/deleted, availability is most important, and the other is relatively small, tons of updates / deletes and compactions, consistency is just as important as availability). It would be great to be able to do so without lots of extra steps after node installation and upgrades or being forced to put storage nodes into chroot or container. Also, this might be useful for running cluster with multiple nodes on same system during development testing (not too sure about that).

For example, let's say we want to run three storage nodes, one from cluster we'll call "main", another one from "temp" and third one without any name (this example is overly complicated to cover both cases of running two nodes that got name and adding a named node to unnamed one). **EDIT**: this ticket is _only_ about multiple instances of nodes that are using global config feature from #791 - it's kind of impossible to run multiple nodes from the same directory with current style of configs that are located inside the same directory.

Logically, we'd want config directories `/etc/leofs/leo_storage`, `/etc/leofs/leo_storage_main` and `/etc/leo_storage_temp` to be used. The problem is how to pass that to current startup script. Let's rely on startup script wrappers from #790 (but maybe there are other options?). This can be solved, for example, if using systemd services for these wrappers: systemd supports templates (so one would name the scripts `leo_storage@.service`, `leo_storage@main.service`, `leo_storage@temp.service`) and that name like "main" or "temp" can be passed to script through environment variable. The fact that "_" and a string should be added to global config directory path is handled in real startup script. Basically, script will contain something like
```
Environment=NODE_EXTRA_NAME=%i
```

However, this would work only for systemd-based systems (EL7, Debian 8/9, Ubuntu 16.04). EL6 and Ubuntu 14.04 are supported as well, it would be nice to have solution for them as well, if possible. Possible ways of doing it include:
1. The wrapper init script that serves as SysV service can be made in multiple copies, e.g. `/etc/init.d/leo_storage` copied to `/etc/init.d/leo_storage_main`. Inside each copy, in header environment variables are defined, e.g.
```
NODE_TYPE=leo_storage
NODE_EXTRA_NAME=main  # or "temp", or empty string
```
which are then read from original scripts
2. Don't write SysV init wrappers at all, as both these systems use upstart, write upstart service config files instead. Not sure how good is this idea (plus I never wrote upstart services, but it shouldn't be too hard, I think). This is variant of previous solution - use would still need to make copies and change them, but upstart services are cleaner than bash scripts and easier for users to edit. Normally this - skipping SysV-init style script and providing upstart one - would seem like a somewhat crazy idea, but given that all system officially supported by LeoFS use either upstart or systemd... why not? This feature doesn't interfere with testing / development anyway, it's for running in real environment, and there isn't much point in providing solution for environments that you aren't supposed to use on real production clusters.

**Problems**:
1.~~The way PIPE_DIR is defined should be changed to allow multiple instances of node to run.~~ **EDIT: not a problem anymore since `foreground_start` launch option for systemd does not use PIPE_DIR.** One way is to add $NODE_EXTA_NAME somewhere (it will stop being kinda-like real path then, but it's OK, I suppose). However, this leaves question on how to use current features of script like "console", "attach" and so on for nodes other than one having empty name. One way is adding an option for that, e.g. `/usr/local/leofs/current/leo_storage/bin/leo_storage -extra-name main attach`. Another is to scratch the idea of systemd / upstart scripts and have init script wrapper that supports passing all these commands. Any other ways? I don't think this feature is needed much on real production, it's mostly for development/debugging, but it would be nice to have a clean solution.
2. While having multiple nodes of the same type running from the same original directory (`/usr/local/leofs/current/leo_storage`) is nice, who would make sure that they are actually completely separated? That involves changing RUNNER_LOG_DIR in environment files and changing *tons* of directories in config files (plus you'll have to copy & change snmp config as well), e.g.
```
mnesia.dir = /var/leofs/leo_manager_0/work/mnesia/127.0.0.1

sasl.sasl_error_log = /var/log/leofs/leo_manager_0/sasl/sasl-error.log
sasl.error_logger_mf_dir = /var/log/leofs/leo_manager_0/sasl

log.erlang = /var/log/leofs/leo_manager_0/erlang
log.app = /var/log/leofs/leo_manager_0/app
log.member_dir = /var/log/leofs/leo_manager_0/ring
log.ring_dir = /var/log/leofs/leo_manager_0/ring

queue_dir = /var/leofs/leo_manager_0/work/queue
erlang.crash_dump = /var/log/leofs/leo_manager_0/erl_crash.dump

snmp_conf = /etc/leofs/leo_manager_0/leo_snmp
```
miss one and you'll have that shared between two instances with possibly disasterous results. Sure, one might say that this feature is for advanced usage and one can expect its users to triple-check everything. But even then: imagine that an extra directory is mentioned in main config file after upgrade and system administrator forgot to re-define it. And other similar problems. However, given that there are so many directories and various ways they work (change to environment variables in script, changes in vm.args, changes in app.config), I'm just not sure how/where to implement this.

Some possible solutions include:
1. Do not manage multiple instance through multiple startup/shutdown script, make a general wrapper which manages starting/stopping all (or individual) instances, checks config files and doesn't allow to launch if some parts of config files point to the same locations. E.g. kind of like postgresql "cluster" is managed in Debian (https://wiki.debian.org/PostgreSql). I'm not a fan of this solution because it makes typical case when you only need a single node of a single type more complicated.
2. Some shared database (in /tmp or something), which is read/written during startup. All directories are written there when node starts, and when node of the same type with different name starts, it checks against that database and if some directory is shared, it gives user the warning and aborts. Problems: while there are lots of ways to make that "shared database" in seemingly reliable way, there are lots of pitfalls like dealing with stale data and locks, requirement to support and update this code as new directories are added and other annoyances. And making a complex solution for a problem that probably doesn't even exist for most users is kind of silly.
3. Changes to config file to allow more global change of directories. It's easy to see in example before that nearly all paths fall into prefix being either `/var/leofs` or `/var/log/leofs`, and the suffix is just like it was in original config. If instead of redefining all these individual paths you could just define two prefixes, it would make the problem simple enough to ignore or implement some very primitive check. Another PITA here is snmp_conf, which you have to copy and put the path to it in main config file just to redefine "db_dir" in snmp config file. I wonder if it might be possible to change that directory through main config file? Idk, maybe having snmp config always generated by cuttlefish and located near generated app.config is possible? I don't see why not, and setting snmp `db_dir` and `config` directly in the main config is hardly any more complicated than setting path to snmp config. Or am I missing something and there was a need to make that config permanent, not generated?

**EDIT**: removed mention of "snmp port" setting. Also I checked problem with snmp settings in more depth and the only difference I can see is that since there are few pre-defined sets of snmp agent configurations with different ports (btw, seem to be made *just* for running multiple nodes of same type on the same system, probably for development, though), with this change you'd need to specify two options pointing to the same number, not one (say, right now it's `snmp/snmpa_gateway_1/leo_gateway_snmp`, and it'll become `snmp/snmpa_gateway_1/db` and `snmp/snmpa_gateway_1/agent/conf`). Which is, of course, a bit more error-prone, but not the end of the world, and is probably avoidable with some cuttlefish template magic.

Anyhow, main problem of third solution is that it probably requires changes that I cannot do myself alone, plus, well, there is no guarantee that it's possible to change it enough to work and not to break any compatibility.

As a last resort option, ignoring the problem - providing such option to run multiple nodes but officially not supporting it, declaring that it's up to users to deal with the problems if they wrote config files poorly.

Ideas? Could there be some easier solution that I've missed?

**Comment by mocchira** [comment]
Thanks for your proposal. WIP, however the solution 3 (Changes to config file to allow more global change of directories) looks promising to me for now and now vetting where we should hack around and how much of work we need.

**Comment by vstax** [comment]
@mocchira 
> the solution 3 (Changes to config file to allow more global change of directories) looks promising to me for now and now vetting where we should hack around and how much of work we need.

I'd like to suggest #838, other than the need to write some code in .schema file to expand a macro-like substitution (I wonder if it's hard to implement in a global way, like, expand all the macros with a single piece of code), I don't think I can see any downsides for this kind of change. Please correct me if I'm wrong...

The proposed change from #837 isn't that nice as it might experience a bit for some users (mostly during developing and testing), I'm not too sure what I'm suggesting there is the best possible way, but that (or similar) solution would simplify current snmp_config annoyance.


Regarding this feature itself. The original idea (check for environment variable like `NODE_EXTRA_NAME=altnode` and change paths to global config from launch script, if it's set) isn't hard to implement, though right now I will be able to support setting it only under systemd. It would work under upstart if/when upstart services are created as well.

I know two ways to implement this in systemd. First of all, launch script is changed (because the feature is actually implemented there). Then, the first solution is - leave current systemd unit files just as they are; add documentation, explaining how user is supposed to make a copy of unit file, e.g.
```
# cp /usr/lib/systemd/system/leofs-storage.service /etc/systemd/system/leofs-storage-alt.service
# vim /etc/systemd/system/leofs-storage-alt.service
(add "Environment=NODE_EXTRA_NAME=altnode" to [Service] section)
```
It's all very simple and obvious for users that don't need this feature, but a bit annoying extra movements for people that do ("annoying" not in sense that it's hard, but in a sense that it's not how it's usually supposed to be done).

Second solution is to use instanced unit, that is, rename service file to `leofs-storage@.service`. Then all is needed to create such node is
```
# systemctl enable leofs-storage@altnode.service
```
The disadvantage here is that all users will have to live that with "@" at the end of service, while running multiple instances is probably required only by a small percent of users. Not to mention that preparing second cluster will require quite a bit more work than just enabling a service (you need to do a proper second set of config files in `/etc/leofs/leo_storage_altnode`, new managers and so on) so simplifying the task of (one-time) service creation seems kind of strange. Of course, instanced units have an advantage of getting the changes that come with new packages, if service files are changed there, they automatically get the new version, while people who had to create copy in /etc/systemd don't.

It's not easy to make the choice between these without knowing how many people exactly would need this feature at all. Generally, first approach (non-instanced service, people who need second instance just make a copy and change it) seems better to me because "@" in the end won't confuse anyone but I might be overthinking this, I don't like having "@" in the end but maybe that's just me. There are no disadvantages of instanced service other than slightly confusing name, it's just that I tend to think that it's quite an annoying disadvantage, so I like the first option more.

**Comment by vstax** [comment]
Trying out this implementation: https://github.com/vstax/leofs/commit/102a10ad525edbd46968ffa569818f76c9d9b808

Running service with `Environment=NODE_EXTRA_NAME=second`:
```
сен 19 21:18:26 stor01.selectel.cloud.lan leo_storage[8811]: Config path: /etc/leofs/leo_storage_second
сен 19 21:18:26 stor01.selectel.cloud.lan leo_storage[8811]: Exec: /usr/local/leofs/current/leo_storage/erts-9.0/bin/erlexec -noinput -boot /usr/local/leofs/current/leo_storage/releases/1/leo_storage -mode minimal -config /etc/leofs/leo_storage_second/app.config -args_file /etc/leofs/leo_storage_second/vm.args -- console
сен 19 21:18:26 stor01.selectel.cloud.lan leo_storage[8811]: Root: /usr/local/leofs/current/leo_storage
```

Besides copy of unit file with that environment variable changed, it required creating `/etc/leofs/leo_storage_second` directory, config files `leo_storage_second/leo_storage.d/20-leo_storage.conf` and `leo_storage/leo_storage.d/20-leo_storage.conf` have different work/log directories, different SNMP config and second node has `rpc.server.listen_port = 13078` set so both of them won't use the same rpc port.

The problem happens when trying to stop the node. Current "stop" code from script - loop around "ps ax" - is not compatible with two nodes running from the same RUNNER_BASE_DIR, as long as one node is still running, that code will never finish.

I'm checking out various solutions. One is to implement #637 and kill the process instead of stopping it with script - however, right now there is no guarantee that this kind of stopping will be reliable, as signals can create various issues, some of which might be hard to work around.

Another solution is to make this code more complicated, ideas I have right now is to use either cgroup information (e.g. get the current cgroup like `grep name= /proc/self/cgroup`, execute something like `ps -ww -eo pid,command:1024,cgroup` and look only at beam.smp in the same cgroup) - this is probably easier but will only work with systemd, not upstart etc.
Alternative solution involves secondary filter to "ps" output and grep for "NODE_EXTRA_NAME=$NODE_EXTRA_NAME"  in `/proc/(beam.smp pid)/environ` for each beam.smp and skip processes that don't match that, but that's a bit ugly. However, it will work with any solution that sets up environment and doesn't depend on systemd's automatically created cgroups.

Other than stopping, I see no issues so far.

**Comment by vstax** [comment]
@mocchira 
I've created cgroup solution https://github.com/leo-project/leofs/commit/189e7d00925383a566e16747785b662fe615c4c6

It looks kind of ugly but actually works fine when working as systemd service (though a hack with stripping "/control" was needed - "ExecStop" is executed in sub-cgroup to main service cgroup). And it works fine when there is no systemd. It works fine when there are no cgroups even (/proc/self/cgroup is empty); it needs changing for very old kernel without cgroup support, of course.

However, only after finishing this experiment I understood that it won't work with launching through script (not service) under systemd because after login-logout new user session will have different cgroup (as systemd creates cgroup for each login session), so stopping won't work.

I don't think I know the way to solve it (I suppose I _could_ ignore cgroups when the script is executed under user.slice as opposed to system.slice but I'm not sure I like how complicated the logic will become), I will probably do a post-filter solution by grepping /proc/$pid/environ instead. What do you think?

**Comment by mocchira** [comment]
@vstax Thank you for digging various options!

First of all, your first solution (with environmental variable) looks promising to me. I think we are on the right track here. 
Regarding the solution for stopping, IMHO the long term solution would be to handle SIGTERM by fixing #637 however that would work only on OTP19 or higher ones and given that now we are supporting R16B03 or higher ones, we also need a short term solution.

> I don't think I know the way to solve it (I suppose I could ignore cgroups when the script is executed under user.slice as opposed to system.slice but I'm not sure I like how complicated the logic will become), I will probably do a post-filter solution by grepping /proc/$pid/environ instead. What do you think?

post-filter by grepping environmental variables through procfs would be good short term solution to me.

**Comment by vstax** [comment]
@mocchira 
Well, to my surprise, ps already supports showing environment variables. Not sure when this feature appeared, but oldest procps I found (3.0.5 from 2002) already supports it, so it should be fine to rely on it.

I was able to come up with relatively clean solution here https://github.com/vstax/leofs/commit/c0136f379c2326b894c2e8cda1314394a503fec1

The only nitpick is that the script relies on the environment variable always being set at launch (to empty value, if nothing else), so when trying to stop, it assumes that process was launched by (new script version) with that environment variable set. If someone upgrades to this version while running node launched by older script, when attempting to stop it with new script it will return right away (because grep won't find that environment variable defined). This is quite minor, I think, as the node will be stopped correctly anyway, and it won't happen after that - and the benefit is code being simpler.

However, the changed code for "stop" command (different ps mode and added grep)  in theory might create some trouble, since this change affects all Linux systems. I mean, I did tests on few various systems (EL6/7, Ubuntu 14/16, Debian 8) but might've missed something.

With #837 and #838 split to separate tickets, the only thing missing here is documentation. I was testing only multiple leo_storage belonging to the same cluster on single system so far, I will submit PR after checking that I can actually run two clusters with both sets of leo_storage working on the same nodes and multiple leo_gateways as well.

I don't think anyone is interested in running leo_managers for two different clusters on the same system, as it doesn't make much sense to me (it should be possible, just will be annoying to both configure that and use leofs-adm); containers or something would work much better here. An interesting note is that this change probably allows to have a single copy of manager node and run master and slave as two instances of that (need to set RUNNER_SCHEMA_DIR= separately in .environment files for that, though) - after all, config and schema are the only things different between them. But, once again, with established release system that creates two copies and works fine like that there isn't much point in this, I think.

**Comment by mocchira** [comment]
@vstax 

> Well, to my surprise, ps already supports showing environment variables. Not sure when this feature appeared, but oldest procps I found (3.0.5 from 2002) already supports it, so it should be fine to rely on it.

Wow I also didn't know that. ps args has been really complicated.. anyway seems more portable solution than retrieving through procfs.

> I was able to come up with relatively clean solution here vstax/leofs@c0136f3

LGTM.

> This is quite minor, I think, as the node will be stopped correctly anyway, and it won't happen after that - and the benefit is code being simpler.

Looks Pros over Cons to me.

> However, the changed code for "stop" command (different ps mode and added grep) in theory might create some trouble, since this change affects all Linux systems. I mean, I did tests on few various systems (EL6/7, Ubuntu 14/16, Debian 8) but might've missed something.

OK. we will do on as much systems as possible.

> With #837 and #838 split to separate tickets, the only thing missing here is documentation. I was testing only multiple leo_storage belonging to the same cluster on single system so far, I will submit PR after checking that I can actually run two clusters with both sets of leo_storage working on the same nodes and multiple leo_gateways as well.

> I don't think anyone is interested in running leo_managers for two different clusters on the same system, as it doesn't make much sense to me (it should be possible, just will be annoying to both configure that and use leofs-adm); containers or something would work much better here. An interesting note is that this change probably allows to have a single copy of manager node and run master and slave as two instances of that (need to set RUNNER_SCHEMA_DIR= separately in .environment files for that, though) - after all, config and schema are the only things different between them. But, once again, with established release system that creates two copies and works fine like that there isn't much point in this, I think.

Got it.

**Comment by vstax** [comment]
@mocchira Fixed a (minor but nasty) bug in handling substrings in environment variables, otherwise it works fine. PR submitted; please review when you have time.

**Comment by vstax** [comment]
@mocchira 
This issue is done for 1.3.8. I don't think there are any problems, but just in case, if you have some time, please test it (by following the corresponding guide in documentation) at least once - probably better on Ubuntu 16.04, because I haven't tested it as through as EL7. There should be no problems with implementation itself but I might've missed / described poorly something in documentation; since I wrote it after implementation and it's hard to "follow" it for me (as I just know how to do it anyway).

**Comment by mocchira** [comment]
@vstax OK. will do.

**Comment by mocchira** [comment]
WIP

**Comment by vstax** [comment]
Looks like there is a new option that I've missed: del_bucket_state_dir (defined in leo_storage.conf and schema)
It's not mentioned in documentation and doesn't seem to be used in code, though? If it will used sometime soon, I suppose it should be added to documentation now just in case (otherwise there will be problems when it's used, at least until #838 is implemented).

**Comment by mocchira** [comment]
@vstax Thanks for pointing out.

> It's not mentioned in documentation and doesn't seem to be used in code, though? If it will used sometime soon, I suppose it should be added to documentation now just in case (otherwise there will be problems when it's used, at least until #838 is implemented).

Yes looks not be used (From what I understand, state for delete-bucket is stored under queue_dir). so it should be no problem.
@yosukehara correct me if there is something wrong. also if it isn't used at all then I'm going to remove those from conf/schema files.

**Comment by yosukehara** [comment]
> del_bucket_state_dir (defined in leo_storage.conf and schema)

@vstax @mocchira I've recognized that we remove it.

**Comment by mocchira** [comment]
@yosukehara Thanks for confirming.

**Comment by mocchira** [comment]
@vstax Done tests on Ubuntu 16.04. It works without any problem if there is no configuration errors.
a little things for improvements to docs as below.

1. at http://leo-project.net/leofs/docs/admin/system_admin/multiple_nodes/#activating-more-instances-of-the-same-node
> Create local copy of systemd unit file with new name: cp /usr/lib/systemd/system/leofs-storage.service /etc/systemd/system/leofs-storage-second.service

As the default location for systemd unit files on Ubuntu 16.04 is `/lib/systemd/system`, this should be mentioned on the doc.

2. http://leo-project.net/leofs/docs/admin/system_admin/multiple_nodes/#list-of-options-that-must-have-different-settings IMO, this section should be placed **before** http://leo-project.net/leofs/docs/admin/system_admin/multiple_nodes/#activating-more-instances-of-the-same-node as the procedure step by step described on `Activating more instances of the same node` would encourage readers to start a systemd service, it would be good to show them the list before starting. (I actually failed to start for the first time and then read the list and succeeded in starting at the second attempt.) 

3. It's obvious however including nodename into the list would be helpful especially for new users.

Please check those out and if you find something to be fixed then let me know or send PR if you have time.

**Comment by vstax** [comment]
@mocchira Thank you for testing. 

> As the default location for systemd unit files on Ubuntu 16.04 is /lib/systemd/system, this should be mentioned on the doc.

Right, I kind of forgot to mention that; I'll fix it

> this section should be placed before http://leo-project.net/leofs/docs/admin/system_admin/multiple_nodes/#activating-more-instances-of-the-same-node

You are right (document was written in assumption that reader will finish it before starting to execute commands, but it should've been better). However, just exchanging sections would break the flow of document, I think. Either a reference to the later section would help, or I need to split the "Activating more instances of the same node" section in the middle (like, move first half to some "Preparing .." section, then list config options, then continue with how to launch). I'll check which way would be better and make a change later.

> It's obvious however including nodename into the list would be helpful especially for new users.

Hm, I didn't write it as requirement because I assumed it is not, at least not always (when nodes belong to different clusters, is it still a requirement? Well I haven't checked if it really works regarding epmd so it might still be). I will add it.

**Comment by mocchira** [comment]
@vstax 

> You are right (document was written in assumption that reader will finish it before starting to execute commands, but it should've been better). However, just exchanging sections would break the flow of document, I think. Either a reference to the later section would help, or I need to split the "Activating more instances of the same node" section in the middle (like, move first half to some "Preparing .." section, then list config options, then continue with how to launch). I'll check which way would be better and make a change later.

Sounds good.

> Hm, I didn't write it as requirement because I assumed it is not, at least not always (when nodes belong to different clusters, is it still a requirement? Well I haven't checked if it really works regarding epmd so it might still be). I will add it.

I see. It might be good to add it with a notice like nodename is required only when nodes belong to the same cluster.
